x/mongo/driver/tranquil_operation.go

```diff
@@ -136,3 +153,374 @@ func (op Operation) decodeOpQuery(wm []byte) (bsoncore.Document, error) {
+			// When running commitTransaction we return WriteConcernErrors as an Error.
+			err := Error{
+				Name:    tt.WriteConcernError.Name,
+				Code:    int32(tt.WriteConcernError.Code),
+				Message: tt.WriteConcernError.Message,
+				Labels:  tt.Labels,
+			}
+			// The UnknownTransactionCommitResult label is added to all writeConcernErrors besides unknownReplWriteConcernCode
+			// and unsatisfiableWriteConcernCode
+			if err.Code != unknownReplWriteConcernCode && err.Code != unsatisfiableWriteConcernCode {
+				err.Labels = append(err.Labels, UnknownTransactionCommitResult)
+			}
+			if retryableErr && setup.retryEnabled {
+				err.Labels = append(err.Labels, RetryableWriteError)
+			}
+			return err, false
+		}
+		operationErr.WriteConcernError = tt.WriteConcernError
+		operationErr.WriteErrors = append(operationErr.WriteErrors, tt.WriteErrors...)
+		operationErr.Labels = tt.Labels
+	case Error:
+		if tt.HasErrorLabel(TransientTransactionError) || tt.HasErrorLabel(UnknownTransactionCommitResult) {
+			op.Client.ClearPinnedServer()
+		}
+		if e := err.(Error); setup.retryable && op.Type == Write && e.UnsupportedStorageEngine() {
+			return ErrUnsupportedStorageEngine, false
+		}
+
+		connDesc := setup.conn.Description()
+		var retryableErr bool
+		if op.Type == Write {
+			retryableErr = tt.RetryableWrite(connDesc.WireVersion)
+			preRetryWriteLabelVersion := connDesc.WireVersion != nil && connDesc.WireVersion.Max < 9
+			// If retryWrites is enabled, add a RetryableWriteError label for network errors and retryable errors from pre-4.4 servers
+			if setup.retryEnabled && (tt.HasErrorLabel(NetworkError) || (retryableErr && preRetryWriteLabelVersion)) {
+				tt.Labels = append(tt.Labels, RetryableWriteError)
+			}
+		} else {
+			retryableErr = tt.RetryableRead()
+		}
+
+		if setup.retryable && retryableErr && setup.retries != 0 {
+			setup.retries--
+			original, err = err, nil
+			setup.conn.Close() // Avoid leaking the connection.
+			setup.srvr, err = op.selectServer(ctx)
+			if err != nil {
+				return original, false
+			}
+			setup.conn, err = setup.srvr.Connection(ctx)
+			if err != nil || setup.conn == nil || !op.retryable(setup.conn.Description()) {
+				if setup.conn != nil {
+					setup.conn.Close()
+				}
+				return original, false
+			}
+			if op.Client != nil && op.Client.Committing {
+				// Apply majority write concern for retries
+				op.Client.UpdateCommitTransactionWriteConcern()
+				op.WriteConcern = op.Client.CurrentWc
+			}
+			return nil, true
+		}
+
+		if op.Client != nil && op.Client.Committing && (retryableErr || tt.Code == 50) {
+			// If we got a retryable error or MaxTimeMSExpired error, we add UnknownTransactionCommitResult.
+			tt.Labels = append(tt.Labels, UnknownTransactionCommitResult)
+		}
+		return tt, false
+	default:
+		return nil, false
+	}
+	return nil, false
+}
+
+func (op Operation) ExecuteDirect(ctx context.Context, scratch []byte) (*wiremessage.MsgHeader, []byte, error) {
```

Comment:
Execute, but without decoding the response. Need to make sure that we decompress and decrypt it though...